File: 20-valid-parentheses/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    const ps = {
        "}":"{",
        ")":"(",
        "]":"["
        };
    for (let char of s) {
        if (char in ps) {
            if (ps[char] !== stack[stack.length - 1]) return false;
            stack.pop();
        } else {
            stack.push(char);
        }
    }
    return stack.length === 0;
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # stk -> can only store open parens
        # if open parens, can store
        # if closed parens, check if most recent val in stk is complement
            # if not false
        # return true if empty stk
        stk = []
        parens_map = {'(': ')', '[': ']', '{': '}'}
        for char in s:
            if char in parens_map:
                stk.append(char)
            elif stk and parens_map.get(stk[-1]) == char:
                stk.pop()
            else:
                return False
        return len(stk) == 0